File: python/leetcode/215.py
import random
class Solution:
    # @param {integer[]} nums
    # @param {integer} k
    # @return {integer}
    def findKthLargest(self, nums, k):
        pivot = random.choice(nums)
        nums1, nums2 = [], []
        for num in nums:
            if num > pivot:
                nums1.append(num)
            elif num < pivot:
                nums2.append(num)
        if k <= len(nums1):
            return self.findKthLargest(nums1, k)
        if k > len(nums) - len(nums2):
            return self.findKthLargest(nums2, k - (len(nums) - len(nums2)))
        return pivot


import heapq
class Solution(object):
    def findKthLargest(self, nums, k):
        h = nums[:k]
        heapq.heapify(h)
        for i in range(k, len(nums)):
            heapq.heappushpop(h, nums[i])
        return heapq.heappop(h)

File: cpp/leetcode/215.cpp
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
     int pivot = rand()%nums.size();
     vector<int> num1, num2;
     for(int i = 0; i < nums.size(); i++){
         if(nums[i] > nums[pivot]) num1.push_back(nums[i]);
         else if (nums[i] < nums[pivot]) num2.push_back(nums[i]);
     }
     if(k <= num1.size()) return findKthLargest(num1, k);
     if(k > nums.size() - num2.size()) return findKthLargest(num2, k - (nums.size() - num2.size()));
     return nums[pivot];
    }
};



class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int> pq(nums.begin(), nums.end());
        for (int i = 0; i < k - 1; i++)
            pq.pop(); 
        return pq.top();
    }
};